### PR TITLE
Okta | Sign In with Email and Password

### PR DESCRIPTION
**Pull request by coldlink**
## What does this change?

Replaces #638 with a cleaned up commit history and updated with existing Okta changes.

The same comments from that PR apply and have been addressed.

Each commit is mostly a clean change and should be possible to follow along that way.

The main focus of this PR is for users to sign in with email and password and set the Okta session cookie along with the IDAPI cookies.

In general the steps are sumarised as follows, assuming okta is enabled, this does not detail all technical requirements, just the main high level:

- User navigates to `/signin`
  - Check `EncryptedState` cookie for `signInRedirect` value
    - if `true` render the sign in page, as we've checked for an existing session
    - if `false` or `undefined` we need to check for an existing session 
      - Check for existing session by performing auth code flow by redirecting user to `/authorize`
      - Okta will redirect to callback endpoint `/oauth/authorization-code/callback`
        - if okta session exists, complete the auth code flow and refreshing identity cookies, redirect to original `returnUrl`
        - if no okta session, okta returns `login_required` error, intercept this
          - set the `EncryptedState` cookie for `signInRedirect` value to `true`
          - redirect to sign in, now that `signInRedirect` is `true`, the page will render
  - If sign in page shown
    - user enters email and password, makes request to gateway `POST /signin`
    - Use okta authenticate endpoint with the email and password `/api/v1/authn`
      - If email/password wrong, show `email or password was incorrect` error
      - If correct, okta returns a single use `sessionToken` which can be exchanged for an session cookie
        - to exchange this for a session cookie, we perform the auth code flow, pass the `sessionToken` to `/authorize` and redirect to okta `/authorize`
          - Okta will redirect to callback endpoint `/oauth/authorization-code/callback`
          - if all good, a session cookie will be set on okta domain, complete the auth code flow and refreshing identity cookies
          - redirect to original `returnUrl`

I've added a number of comments where appropriate to add reasoning to why things are in a particular way.

## Other change
I refactored the `crypto` library that we used to encrypt/decrypt the `EncryptedState` cookie (https://github.com/guardian/gateway/pull/1410/commits/45b202b8c896e58ade2cf970396c9a8bea127628). This removes the requirement of `getConfiguration` being needed by those methods, in turn meaning we're able to use the module outside of the server, specifically for cypress where we're now able to test and set the encrypted state from cypress tests (https://github.com/guardian/gateway/pull/1410/commits/72f12649e5ec4b830d93b2b7fc3f04eb85a7f43c).

#### Authentication API
https://developer.okta.com/docs/reference/api/authn/#primary-authentication-with-public-application

#### OAuth 2.0 and OpenID connect documentation
https://developer.okta.com/docs/concepts/oauth-openid/

#### Retrieving a session cookie
https://developer.okta.com/docs/guides/session-cookie/main/#retrieve-a-session-cookie-through-the-openid-connect-authorization-endpoint

#### Authorization code flow
https://developer.okta.com/docs/concepts/oauth-openid/#authorization-code-flow

#### `/authorize` endpoint
https://developer.okta.com/docs/reference/api/oidc/#authorize

## Main changes
- [x] Implement  certified OIDC Library - [`openid-client`](https://github.com/panva/node-openid-client) - https://github.com/guardian/gateway/pull/1410/commits/fe53f01b1bbb937acbfb68d761fbc1c834183e45
- [x] Implement Authorization Code flow - https://github.com/guardian/gateway/pull/1410/commits/9303ea8acdcff1ed16624edd7ecf985b94aecb16
  - [x] Redirect to Okta `/authorize` after authentication
  - [x] Implement `redirect_uri` endpoint which handles the return from okta `/authorize`
  - [x] Exchange auth code for access token
  - [x] Redirect to `return_url` once authenticated
- [x] Identity implementation to retrieve identity cookies - https://github.com/guardian/gateway/pull/1410/commits/976e897e489bd49fc3783801361ca2efd8b474f8
- [x] Change existing Okta authenticate endpoint - https://github.com/guardian/gateway/pull/1410/commits/96583196e05ddff0682cc5fd8d1169cb4f4b91d9
- [x] Add `POST /signin` support for okta - https://github.com/guardian/gateway/pull/1410/commits/9b03a8a1bb64b8b06e8d77a7eb9d52268d653c2c
- [x] Existing session check before showing sign in page - https://github.com/guardian/gateway/pull/1410/commits/f08552b478443ad3b0c1ff8c92c1135bbfa08145
- [x] Error handling
- [x] Add cypress tests if possible (at least some mocked tests)
- [x] Additional comments/documentation

## Flow Diagram
This PR focuses on the steps between `Browser`|`Gateway`|`Okta`|`Identity`. The bit with `theguardian.com` will be worked on when we start migrating individual apps.

Guardian_Sign_In_from_theguardian com

Flow for when a preexisting session does not exist:

122358182-50809400-cf4c-11eb-9f9c-1d7b4389bd61

### DEV
This PR makes things easier: https://github.com/guardian/identity-platform/pull/495
- [x] Pull down the correct `.env` for the environment your testing at
  - The Okta environment and the IDAPI environment must match
- [x] Launch gateway, navigatge to `/signin?useOkta=true`, the flag is required
- [x] Attempt to sign in
- [x] Go back to `/signin?useOkta=true`, you should be redirected to the homepage
  - This is the check to see if the session exists
- [x] Clear the cookies on the relevant okta and identity domains to test `/signin` again

### CODE
- [x] Enable the feature switch on `CODE`
- [x] Deploy to `CODE`
- [x] Update VCL to add /oauth route -> https://github.com/guardian/identity-platform/pull/496
- [x] Make sure to test sign in routes and states with users who exists in both Identity DB and Okta
- [x] Navigatge to `/signin?useOkta=true`, the flag is required
- [x] Attempt to sign in
- [x] Go back to `/signin?useOkta=true`, you should be redirected to the homepage
  - This is the check to see if the session exists
- [x] Clear the cookies on the relevant okta and identity domains to test `/signin` again

### CODE REVIEW

**Review by ob6160 — COMMENTED** [review]
this is really great work 💯 , looks really solid to me. Two comments from my first read through, I'll take another look tomorrow with a fresh pair of eyes as well.

One question about end to end testing, what's the plan with that side of things? I'm guessing there are a few blockers to getting them up and running

File: src/server/routes/oauth.ts [hunk 1]
```diff
@@ -0,0 +1,178 @@
...
+      const callbackParams = ProfileOpenIdClient.callbackParams(req);
+
+      // get the oidc_auth_state cookie which contains the "nonce"
+      // and "returnUrl" so we can get the user back to the page they
+      // initially landed on sign in from
+      const authState = getAuthorizationStateCookie(req);
+
+      // check if the state cookie exists
+      if (!authState) {
+        // no state! is this an attack?
```
[the remark this answers is not in the document]
Comment (coldlink) [reply src/server/routes/oauth.ts:87]: the `authState` would have to be set by gateway before performing the redirect to `/authorize`, which happens here: https://github.com/guardian/gateway/pull/1410/files#diff-88981251923a306ae51df84b8db7eae8eab850112e93512e07c21fc7c151dccbR54

If this doesn't exist, that would mean that either a) the state isn't being set correctly, or b) someone is trying to attack the oauth flow for example with an invalid state cookie, or without the state cookie

basically the state parameter is used to prevent CSRF attacks, if the state parameter is missing, or invalid, or different it could mean that someone is trying those kinds of attacks.

i'll update the comments

File: src/server/lib/okta/api/errors.ts [hunk 2]
```diff
@@ -11,8 +11,33 @@ import {
...
+      return {
+        errorCode: error.errorCode,
+        errorSummary: error.errorSummary,
+        errorLink: error.errorLink,
+        errorId: error.errorId,
+        errorCauses: error.errorCauses,
+      };
+    });
+  } catch (error) {
+    throw new OktaAPIResponseParsingError(`${error}`);
```
[the remark this answers is not in the document]
Comment (coldlink) [reply src/server/lib/okta/api/errors.ts:33]: good shout, i'll refactor the implementation of `OktaAPIResponseParsingError` to check the `typeof error` and handle that way rather than just casting to string

this will only be used in edge cases when the response from okta fails to get parsed

[the remark this answers is not in the document]
Comment (ob6160) [reply src/server/lib/okta/api/errors.ts]: sounds good

File: src/server/routes/oauth.ts (hunk 1 above)
[the remark this answers is not in the document]
Comment (ob6160) [reply src/server/routes/oauth.ts]: thanks, the extra info is really helpful :)

File: src/server/lib/okta/openid-connect.ts [hunk 3]
```diff
@@ -0,0 +1,242 @@
...
+import { getConfiguration } from '@/server/lib/getConfiguration';
+import { getProfileUrl } from '@/server/lib/getProfileUrl';
+import { ResponseWithRequestState } from '@/server/models/Express';
+import { logger } from '@/server/lib/serverSideLogger';
+import { RoutePaths } from '@/shared/model/Routes';
+
+/**
+ * An object to hold the state the user was in while performing the
+ * Authorization Code flow.
+ * - `nonce` - Used as the `state` parameter in the Authorization Code Flow
```
Comment (dominickendrick) [inline src/server/lib/okta/openid-connect.ts:16]: Is it worth linking out to the docs that explain the `state` param ? (https://developer.okta.com/docs/reference/api/oidc/#parameter-details)

File: src/server/models/Metrics.ts [hunk 4]
```diff
@@ -2,36 +2,37 @@
...
   | 'OktaRegistrationResendEmail'
   | 'OktaSetPasswordOnWelcomePage'
-  | 'OktaWelcomeResendEmail';
+  | 'OktaWelcomeResendEmail'
+  | 'Register'
+  | 'SendMagicLink'
+  | 'SendValidationEmail'
+  | 'SignIn'
+  | 'UpdatePassword'
+  | `RecaptchaMiddleware`;
```
Comment (dominickendrick) [inline src/server/models/Metrics.ts:29]: single quotes rather than backticks ?

File: src/server/routes/oauth.ts [hunk 5]
```diff
@@ -0,0 +1,185 @@
...
+ * back to the client. Be sure to log the error though!
+ */
+const redirectForGenericError = (
+  req: Request,
+  res: ResponseWithRequestState,
+) => {
+  // we set the sign in redirect to true so that we can render the sign in page
+  // without getting into a redirect loop
+  const encryptedState = readEncryptedStateCookie(req);
+  setEncryptedStateCookie(res, {
```
Comment (dominickendrick) [inline src/server/routes/oauth.ts:56]: Is it worth updating these two calls to be `updateEcryptedStateCookie(req: Request, res: ResponseWithRequestState ,state: EncryptedState)` and you can perform the  `readEncryptedStateCookie` inside the update function ?

File: src/server/routes/signIn.ts [hunk 6]
```diff
@@ -15,84 +15,240 @@ import { ApiError } from '@/server/models/Error';
...
 
-    return res.redirect(303, returnUrl || defaultReturnUri);
-  }),
-);
+    // we now need to generate an okta session
+    // so we'll call the OIDC /authorize endpoint which sets a session cookie
+    // we'll pretty much be performing the Authorization Code Flow
+    return performAuthCodeFlow(res, response.sessionToken);
+  } catch (error) {
+    trackMetric('SignIn::Failure');
```
Comment (dominickendrick) [inline src/server/routes/signIn.ts:223]: Could we have a sepereate metric for Okta ? ie `SignInOkta::Failure` ?

Comment (coldlink) [reply src/server/routes/signIn.ts]: I also made the `trackMetic` method return `void` to make sure that it's explicitly defined as a fire and forget method.

**Review by rebecca-thompson — COMMENTED** [review]
Looks really good! My main question is around our use of the `state` parameter to prevent CSRF attacks and whether to include the `nonce` parameter to prevent replay attacks (not strictly necessary in the auth code flow, but worth a conversation).

File: src/server/lib/okta/openid-connect.ts [hunk 7]
```diff
@@ -0,0 +1,252 @@
...
+  TEMPORARILY_UNAVAILABLE = 'temporarily_unavailable',
+  INVALID_CLIENT = 'invalid_client',
+  LOGIN_REQUIRED = 'login_required',
+  INVALID_REQUEST = 'invalid_request',
+  USER_CANCELED_REQUEST = 'user_canceled_request',
+}
+
+/**
+ * Generate a cryptographically secure random string to be used
+ * as the `state` parameter in the authorization code flow
```
[the remark this answers is not in the document]
Comment (coldlink) [reply src/server/lib/okta/openid-connect.ts:124]: After our discussion, I've renamed our usages of `nonce` to `stateParam` to make it more clear about it's usage, and not to confuse it with the other `nonce` parameter

File: src/shared/lib/featureSwitches.ts [hunk 8]
```diff
@@ -17,8 +17,8 @@ interface FeatureSwitches {
 
 export const featureSwitches: FeatureSwitches = {
   demoSwitch: false,
-  oktaRegistrationEnabled: {
-    DEV: false,
+  oktaEnabled: {
+    DEV: true,
```
[the remark this answers is not in the document]
Comment (coldlink) [reply src/shared/lib/featureSwitches.ts:21]: As discussed as standup, we're okay with leaving this on now, as current usage still requires the use of the `useOkta=true` query parameter

File: cypress/integration/mocked/sign_in.spec.ts [hunk 9]
```diff
@@ -145,6 +145,151 @@ describe('Sign in flow', () => {
...
+    beforeEach(() => {
+      // set the encrypted state signInRedirect to true to render sign in page
+      // we can't mock the session check because it's a redirect to and from okta
+      cy.setEncryptedStateCookie({ signInRedirect: true });
+    });
+
+    it('renders global error if there is the error_description parameter in url', function () {
+      const error = 'There has been an error';
+      cy.visit(
+        `/signin?error_description=${encodeURIComponent(error)}&useOkta=true`,
```
[the remark this answers is not in the document]
Comment (coldlink) [reply cypress/integration/mocked/sign_in.spec.ts:160]: As discussed in standup, we'll propagate a generic error for now until we know how to properly deal with the `error_description` parameter from okta

**Review by rebecca-thompson — APPROVED** [review]

**Review by dominickendrick — APPROVED** [review]

File: src/server/lib/okta/openid-connect.ts [hunk 10]
```diff
@@ -0,0 +1,252 @@
...
+ * @type {string}
+ */
+const AuthorizationStateCookieName = 'GU_oidc_auth_state';
+
+/**
+ * Authorization state cookie options
+ * No expiry/max-age, so cookie is deleted when browser session closes
+ */
+const AuthorizationStateCookieOptions: CookieOptions = {
+  httpOnly: true,
```
Comment (ob6160) [inline src/server/lib/okta/openid-connect.ts:169]: In `encryptedStateCookieOptions`, these boolean values are set using: `!baseUri.includes('localhost')`, might be completely unnecessary but just wanted to double check that this doesn't need the same

File: src/server/lib/okta/openid-connect.ts [hunk 11]
```diff
@@ -0,0 +1,252 @@
...
+  res: ResponseWithRequestState,
+) => {
+  try {
+    res.cookie(
+      AuthorizationStateCookieName,
+      Buffer.from(JSON.stringify(state)).toString('base64'),
+      AuthorizationStateCookieOptions,
+    );
+  } catch (error) {
+    logger.error(`setAuthorizationStateCookie Error`, error);
```
Comment (ob6160) [inline src/server/lib/okta/openid-connect.ts:199]:
```suggestion
    logger.error('setAuthorizationStateCookie Error', error);
```

File: src/server/lib/okta/openid-connect.ts [hunk 12]
```diff
@@ -0,0 +1,252 @@
...
+      Buffer.from(stateCookie, 'base64').toString('utf-8'),
+    );
+
+    if (isAuthorizationState(parsed)) {
+      return parsed;
+    }
+
+    return null;
+  } catch (error) {
+    logger.error(`getAuthorizationStateCookie Error`, error);
```
Comment (ob6160) [inline src/server/lib/okta/openid-connect.ts:231]:
```suggestion
    logger.error('getAuthorizationStateCookie Error', error);
```

File: src/server/routes/signIn.ts [hunk 13]
```diff
@@ -15,84 +15,254 @@ import { ApiError } from '@/server/models/Error';
...
+    // we now need to generate an okta session
+    // so we'll call the OIDC /authorize endpoint which sets a session cookie
+    // we'll pretty much be performing the Authorization Code Flow
+    return performAuthCodeFlow(res, response.sessionToken);
+  } catch (error) {
+    trackMetric('OktaSignIn::Failure');
+
+    logger.error('Okta authentication error:', error);
+
+    const { message, status } = ((error: unknown) => {
```
Comment (ob6160) [inline src/server/routes/signIn.ts:241]: could this function have a name for extra clarity perhaps?

**Review by ob6160 — APPROVED** [review]
:shipit: few small things, but nothing blocking

File: src/server/lib/okta/openid-connect.ts (hunk 10 above)
Comment (coldlink) [reply src/server/lib/okta/openid-connect.ts:169]: ah good shout, I'll update the options to use the localhost check, otherwise things won't work on that domain as expected

File: src/server/routes/signIn.ts (hunk 13 above)
Comment (coldlink) [reply src/server/routes/signIn.ts:241]: i'll refactor this out into it's own method!
